std/algorithm/comparison.d

```diff
@@ -913,81 +912,72 @@ template equal(alias pred = "a == b")
         for element, according to binary predicate `pred`.
     +/
     bool equal(Range1, Range2)(Range1 r1, Range2 r2)
-    if (!useCodePoint!(Range1, Range2) &&
-        isInputRange!Range1 && isInputRange!Range2 &&
+    if (isInputRange!Range1 && isInputRange!Range2 &&
+        !(isInfinite!Range1 && isInfinite!Range2) &&
         is(typeof(binaryFun!pred(r1.front, r2.front))))
     {
-        static assert(!(isInfinite!Range1 && isInfinite!Range2),
-            "Both ranges are known to be infinite");
-
-        //No pred calls necessary
-        static if (isEmptyRange!Range1 || isEmptyRange!Range2)
-        {
-            return r1.empty && r2.empty;
-        }
-        else static if ((isInfinite!Range1 && hasFixedLength!Range2) ||
-            (hasFixedLength!Range1 && isInfinite!Range2))
-        {
-            return false;
-        }
-        //Detect default pred and compatible dynamic array
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isArray!Range1 && isArray!Range2 && is(typeof(r1 == r2)))
+        static if (!useCodePoint!(Range1, Range2))
```

Comment:
By combining the functions, we don't need `useCodePoint` outside this function (it was only used to distinguish the overload). However, there is a slight difference here, as the constraint has checked the predicate against the range type, but in the case of two differing character widths, the code is going to reroute it to comparing dchars instead. Given that the vast vast majority (possibly unanimous) usage of comparing ranges of characters is going to rely on the default comparison of a == b, it's probably not an issue.

Reply:
Oh wait that wasn't intended. Can you explain a bit more? So the old code had these overloads:

```d
bool equal(Range1, Range2)(Range1 r1, Range2 r2)
if (!useCodePoint!(Range1, Range2) &&
    isInputRange!Range1 && isInputRange!Range2 &&
    is(typeof(binaryFun!pred(r1.front, r2.front))));
bool equal(Range1, Range2)(Range1 r1, Range2 r2)	
    if (useCodePoint!(Range1, Range2));
```

My intent was simply to merge these into one while preserving the semantics:

```d
bool equal(Range1, Range2)(Range1 r1, Range2 r2)
if (isInputRange!Range1 && isInputRange!Range2 &&
    is(typeof(binaryFun!pred(r1.front, r2.front))));
```

and use `static if` inside to discriminate on `useCodePoint!(Range1, Range2`. Where did I go wrong?

Reply:
What I mean is effectively prior to this change, you had this constraint on the second form:

```D
if (useCodePoint!(Range1, Range2))
```

And now the constraint on the second form is:

```D
if(useCodePoint!(Range1, Range2) && // this first part is embedded inside the function
    isInputRange!Range1 && isInputRange!Range2 &&
    is(typeof(binaryFun!pred(r1.front, r2.front))));
```

So suppose you pass in a range of `char`, and a range of `wchar`. This is going to use the "decode-to-dchar" mechanism, but the constraint is checking if `binaryFun!pred(char, wchar)` compiles even though that is never actually called.

Again, I think it's a nonsensical case -- who is going to use the function called `equal` to check for something other than equality? Most people who want a transformation would use map to first transform the elements, and then use `equal` on that. But there is that slight difference.

If we wanted it to be "correct", I would say leave the secondary form separate. And in fact, the secondary form technically should check that the predicate works on `(dchar, dchar)` in that case. I left it out because it was a simple wrapper, and I knew it would be checked in the forwarded call properly.

Reply:
In theory, `pred` does not need to be strict equality; it can implement an equivalence relation.  For example, one might want to compare two strings in a case-insensitive manner (e.g., 'A' and 'a' are in the same equivalence class, and so forth).

Reply:
Right, but the question is, will evaluating whether the predicate can be called with the individual char types should affect being able to call it.

I mean, I can't imagine a case where you can call the predicate using dchar, but not using char or wchar.

Reply:
Got it, will fix just to preserve existing behavior. Thanks @schveiguy !

Reply:
I woke up with the feeling we have a bad solution. First off, the documentation will show the private symbol `useCodePoint`. That could be fixed with some awkwardness, but then there's the simple truth that `useCodePoint` is a sheer implementation choice that the user does not want or need to know about.

I'll work out something else - perhaps move the predicate validation to a `static assert`.

Reply:
I opted for pushing `useCodePoint` all the way in and leaving the predicate in the condition. The conditions under that wouldn't compile are vanishingly rare.

Reply:
Yeah, I agree. Thanks!